api/app.go

```diff
@@ -0,0 +1,295 @@
+		c.SetInvalidParam("registerOAuthApp", "app")
+		return
+	}
+
+	secret := model.NewId()
+
+	app.ClientSecret = secret
+	app.CreatorId = c.Session.UserId
+
+	if result := <-Srv.Store.App().Save(app); result.Err != nil {
+		c.Err = result.Err
+		return
+	} else {
+		app = result.Data.(*model.App)
+		app.ClientSecret = secret
+
+		c.LogAudit("client_id=" + app.Id)
+
+		w.Write([]byte(app.ToJson()))
+		return
+	}
+
+}
+
+func allowOAuth(c *Context, w http.ResponseWriter, r *http.Request) {
+	c.LogAudit("attempt")
+
+	w.Header().Set("Content-Type", "application/x-www-form-urlencoded")
+	responseData := map[string]string{}
+
+	responseType := r.URL.Query().Get("response_type")
+	if len(responseType) == 0 {
+		c.Err = model.NewOAuthAppError("allowOAuth", "invalid_request", "Bad response_type")
+		return
+	}
+
+	clientId := r.URL.Query().Get("client_id")
+	if len(clientId) != 26 {
+		c.Err = model.NewOAuthAppError("allowOAuth", "invalid_request", "Bad client_id")
+		return
+	}
+
+	redirectUri := r.URL.Query().Get("redirect_uri")
+	scope := r.URL.Query().Get("scope")
+	state := r.URL.Query().Get("state")
+
+	var app *model.App
+	if result := <-Srv.Store.App().Get(clientId); result.Err != nil {
+		c.Err = model.NewOAuthAppError("allowOAuth", "server_error", "Error accessing the database")
+		return
+	} else {
+		app = result.Data.(*model.App)
+	}
+
+	callback := app.CallbackUrl
+
+	if len(redirectUri) > 0 {
+		registeredUrl, _ := url.Parse(app.CallbackUrl)
+		redirectUrl, _ := url.Parse(redirectUri)
+
+		if redirectUrl.Scheme != "https" {
+			c.Err = model.NewOAuthAppError("allowOAuth", "invalid_request", "Supplied redirect_uri must be https")
+			return
+		}
+
+		if registeredUrl.Host != redirectUrl.Host {
+			c.Err = model.NewOAuthAppError("allowOAuth", "invalid_request", "Supplied redirect_uri host/port did not match registered callback_url")
+			return
+		}
+
+		callback = redirectUri
+	}
+
+	if responseType != model.AUTHCODE_RESPONSE_TYPE {
+		responseData["redirect"] = callback + "?error=unsupported_response_type&state=" + state
+		w.Write([]byte(model.MapToJson(responseData)))
+		return
+	}
+
+	authData := &model.AuthData{UserId: c.Session.UserId, ClientId: clientId, CreateAt: model.GetMillis(), RedirectUri: redirectUri, State: state, Scope: scope}
+	authData.Code = model.HashPassword(fmt.Sprintf("%v:%v:%v:%v", clientId, redirectUri, authData.CreateAt, c.Session.UserId))
+
+	if result := <-Srv.Store.AuthData().Save(authData); result.Err != nil {
+		responseData["redirect"] = callback + "?error=server_error&state=" + state
+		w.Write([]byte(model.MapToJson(responseData)))
+		return
+	}
+
+	authCodeCache.Add(authData.Code, authData)
+	c.LogAudit("success")
+
+	responseData["redirect"] = callback + "?code=" + url.QueryEscape(authData.Code) + "&state=" + url.QueryEscape(authData.State)
+
+	w.Write([]byte(model.MapToJson(responseData)))
+}
+
+func getAccessToken(c *Context, w http.ResponseWriter, r *http.Request) {
+	c.LogAudit("attempt")
```

Comment:
I didn't see any failed audit msgs?  not sure if we want to be extra anal and call them out.